man/RhumbSolve.1
.\" -*- mode: troff; coding: utf-8 -*-
.\" Automatically generated by Pod::Man v6.0.2 (Pod::Simple 3.45)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" \*(C` and \*(C' are quotes in nroff, nothing in troff, for use with C<>.
.ie n \{\
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Required to disable full justification in groff 1.23.0.
.if n .ds AD l
.\" ========================================================================
.\"
.IX Title "RHUMBSOLVE 1"
.TH RHUMBSOLVE 1 2025-11-05 "GeographicLib 2.7" "GeographicLib Utilities"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH NAME
RhumbSolve \-\- perform rhumb line calculations
.SH SYNOPSIS
.IX Header "SYNOPSIS"
\&\fBRhumbSolve\fR [ \fB\-i\fR | \fB\-L\fR \fIlat1\fR \fIlon1\fR \fIazi12\fR ]
[ \fB\-e\fR \fIa\fR \fIf\fR ] [ \fB\-u\fR ]
[ \fB\-d\fR | \fB\-:\fR ] [ \fB\-w\fR ] [ \fB\-p\fR \fIprec\fR ] [ \fB\-E\fR ]
[ \fB\-\-comment\-delimiter\fR \fIcommentdelim\fR ]
[ \fB\-\-version\fR | \fB\-h\fR | \fB\-\-help\fR ]
[ \fB\-\-input\-file\fR \fIinfile\fR | \fB\-\-input\-string\fR \fIinstring\fR ]
[ \fB\-\-line\-separator\fR \fIlinesep\fR ]
[ \fB\-\-output\-file\fR \fIoutfile\fR ]
.SH DESCRIPTION
.IX Header "DESCRIPTION"
The path with constant heading between two points on the ellipsoid at
(\fIlat1\fR, \fIlon1\fR) and (\fIlat2\fR, \fIlon2\fR) is called the rhumb line or
loxodrome.  Its length is \fIs12\fR and the rhumb line has a forward
azimuth \fIazi12\fR along its length.  The quantity \fIS12\fR is the area
between the rhumb line from point 1 to point 2 and the equator; i.e.,
it is the area, measured counter\-clockwise, of the geodesic
quadrilateral with corners (\fIlat1\fR,\fIlon1\fR), (0,\fIlon1\fR),
(0,\fIlon2\fR), and (\fIlat2\fR,\fIlon2\fR).  The longitude becomes
indeterminate when a rhumb line passes through a pole, and
\&\fBRhumbSolve\fR reports NaNs for the longitude and the area in this
case.
.PP
\&\fBNOTE:\fR the rhumb line is \fBnot\fR the shortest path between two points;
that is the geodesic and it is calculated by \fBGeodSolve\fR\|(1).
.PP
\&\fBRhumbSolve\fR operates in one of three modes:
.IP 1. 4
By default, \fBRhumbSolve\fR accepts lines on the standard input containing
\&\fIlat1\fR \fIlon1\fR \fIazi12\fR \fIs12\fR and prints \fIlat2\fR \fIlon2\fR \fIS12\fR on
standard output.  This is the direct calculation.
.IP 2. 4
With the \fB\-i\fR option, \fBRhumbSolve\fR performs the inverse calculation.
It reads lines containing \fIlat1\fR \fIlon1\fR \fIlat2\fR \fIlon2\fR and prints
the values of \fIazi12\fR \fIs12\fR \fIS12\fR for the corresponding shortest
rhumb lines.
.IP 3. 4
Command line arguments \fB\-L\fR \fIlat1\fR \fIlon1\fR \fIazi12\fR specify a rhumb
line.  \fBRhumbSolve\fR then accepts a sequence of \fIs12\fR values (one per
line) on standard input and prints \fIlat2\fR \fIlon2\fR \fIS12\fR for each.
This generates a sequence of points on a rhumb line.
.SH OPTIONS
.IX Header "OPTIONS"
.IP \fB\-i\fR 4
.IX Item "-i"
perform an inverse calculation (see 2 above).
.IP "\fB\-L\fR \fIlat1\fR \fIlon1\fR \fIazi12\fR" 4
.IX Item "-L lat1 lon1 azi12"
line mode (see 3 above); generate a sequence of points along the rhumb
line specified by \fIlat1\fR \fIlon1\fR \fIazi12\fR.  The \fB\-w\fR flag can be used
to swap the default order of the 2 geographic coordinates, provided that
it appears before \fB\-L\fR.
.IP "\fB\-e\fR \fIa\fR \fIf\fR" 4
.IX Item "-e a f"
specify the ellipsoid via the equatorial radius, \fIa\fR and
the flattening, \fIf\fR.  Setting \fIf\fR = 0 results in a sphere.  Specify
\&\fIf\fR < 0 for a prolate ellipsoid.  A simple fraction, e.g., 1/297,
is allowed for \fIf\fR.  By default, the WGS84 ellipsoid is used, \fIa\fR =
6378137 m, \fIf\fR = 1/298.257223563.
.IP \fB\-u\fR 4
.IX Item "-u"
unroll the longitude.  Normally, on output longitudes are reduced to lie
in [\-180deg,180deg).  However with this option, the returned longitude
\&\fIlon2\fR is "unrolled" so that \fIlon2\fR \- \fIlon1\fR indicates how often and
in what sense the geodesic has encircled the earth.
.IP \fB\-d\fR 4
.IX Item "-d"
output angles as degrees, minutes, seconds instead of decimal degrees.
.IP \fB\-:\fR 4
.IX Item "-:"
like \fB\-d\fR, except use : as a separator instead of the d, \*(Aq, and "
delimiters.
.IP \fB\-w\fR 4
.IX Item "-w"
on input and output, longitude precedes latitude (except that on input
this can be overridden by a hemisphere designator, \fIN\fR, \fIS\fR, \fIE\fR,
\&\fIW\fR).
.IP "\fB\-p\fR \fIprec\fR" 4
.IX Item "-p prec"
set the output precision to \fIprec\fR (default 3); \fIprec\fR is the
precision relative to 1 m.  See "PRECISION".
.IP \fB\-E\fR 4
.IX Item "-E"
By default, the rhumb line calculations are carried out using series
expansions valid for |\fIf\fR| < 0.01.  If \fB\-E\fR is supplied, exact
equations for the rhumb line are used and the area integral is
computed with an accurate fit based on this exact equations; these are
valid for arbitrary eccentricities.
.IP "\fB\-\-comment\-delimiter\fR \fIcommentdelim\fR" 4
.IX Item "--comment-delimiter commentdelim"
set the comment delimiter to \fIcommentdelim\fR (e.g., "#" or "//").  If
set, the input lines will be scanned for this delimiter and, if found,
the delimiter and the rest of the line will be removed prior to
processing and subsequently appended to the output line (separated by a
space).
.IP \fB\-\-version\fR 4
.IX Item "--version"
print version and exit.
.IP \fB\-h\fR 4
.IX Item "-h"
print usage and exit.
.IP \fB\-\-help\fR 4
.IX Item "--help"
print full documentation and exit.
.IP "\fB\-\-input\-file\fR \fIinfile\fR" 4
.IX Item "--input-file infile"
read input from the file \fIinfile\fR instead of from standard input; a file
name of "\-" stands for standard input.
.IP "\fB\-\-input\-string\fR \fIinstring\fR" 4
.IX Item "--input-string instring"
read input from the string \fIinstring\fR instead of from standard input.
All occurrences of the line separator character (default is a semicolon)
in \fIinstring\fR are converted to newlines before the reading begins.
.IP "\fB\-\-line\-separator\fR \fIlinesep\fR" 4
.IX Item "--line-separator linesep"
set the line separator character to \fIlinesep\fR.  By default this is a
semicolon.
.IP "\fB\-\-output\-file\fR \fIoutfile\fR" 4
.IX Item "--output-file outfile"
write output to the file \fIoutfile\fR instead of to standard output; a
file name of "\-" stands for standard output.
.SH INPUT
.IX Header "INPUT"
\&\fBRhumbSolve\fR measures all angles in degrees, all lengths (\fIs12\fR) in
meters, and all areas (\fIS12\fR) in meters^2.  On input angles (latitude,
longitude, azimuth, arc length) can be as decimal degrees or degrees,
minutes, seconds.  For example, \f(CW\*(C`40d30\*(C'\fR, \f(CW\*(C`40d30\*(Aq\*(C'\fR, \f(CW\*(C`40:30\*(C'\fR, \f(CW\*(C`40.5d\*(C'\fR,
and \f(CW40.5\fR are all equivalent.  By default, latitude precedes longitude
for each point (the \fB\-w\fR flag switches this convention); however on
input either may be given first by appending (or prepending) \fIN\fR or
\&\fIS\fR to the latitude and \fIE\fR or \fIW\fR to the longitude.  Azimuths are
measured clockwise from north; however this may be overridden with \fIE\fR
or \fIW\fR.
.PP
For details on the allowed formats for angles, see the \f(CW\*(C`GEOGRAPHIC
COORDINATES\*(C'\fR section of \fBGeoConvert\fR\|(1).
.SH PRECISION
.IX Header "PRECISION"
\&\fIprec\fR gives precision of the output with \fIprec\fR = 0 giving 1 m
precision, \fIprec\fR = 3 giving 1 mm precision, etc.  \fIprec\fR is the
number of digits after the decimal point for lengths.  For decimal
degrees, the number of digits after the decimal point is \fIprec\fR + 5.
For DMS (degree, minute, seconds) output, the number of digits after the
decimal point in the seconds component is \fIprec\fR + 1.  The minimum
value of \fIprec\fR is 0 and the maximum is 10.
.SH ERRORS
.IX Header "ERRORS"
An illegal line of input will print an error message to standard output
beginning with \f(CW\*(C`ERROR:\*(C'\fR and causes \fBRhumbSolve\fR to return an exit code
of 1.  However, an error does not cause \fBRhumbSolve\fR to terminate;
following lines will be converted.
.SH ACCURACY
.IX Header "ACCURACY"
The algorithm used by \fBRhumbSolve\fR uses either series expansions or
(if \fB\-E\fR is specified) exact formulas for computing the rhumb line
and the area.  These series are formulas are accurate for |\fIf\fR| <
0.01 and the exact formulas apply for any value of the flattening.
The computation of rhumb lines and the area involves the ratio of
differences and, for nearly east\- or west\-going rhumb lines, this
might result in a large loss of accuracy.  However, this problem is
avoided by the use of divided differences. For the WGS84 ellipsoid,
the error is about 10 nanometers using either method.
.SH EXAMPLES
.IX Header "EXAMPLES"
Route from JFK Airport to Singapore Changi Airport:
.PP
.Vb 2
\&   echo 40:38:23N 073:46:44W 01:21:33N 103:59:22E |
\&   RhumbSolve \-i \-: \-p 0
\&
\&   103:34:58.2 18523563 45921660958919
.Ve
.PP
N.B. This is \fBnot\fR the route typically taken by aircraft because it\*(Aqs
considerably longer than the geodesic given by \fBGeodSolve\fR\|(1).
.PP
Waypoints on the route at intervals of 2000km:
.PP
.Vb 2
\&   for ((i = 0; i <= 20; i += 2)); do echo ${i}000000;done |
\&   RhumbSolve \-L 40:38:23N 073:46:44W 103:34:58.2 \-: \-p 0
\&
\&   40:38:23.0N 073:46:44.0W 0
\&   36:24:30.3N 051:28:26.4W 9817078307821
\&   32:10:26.8N 030:20:57.3W 18224745682005
\&   27:56:13.2N 010:10:54.2W 25358020327741
\&   23:41:50.1N 009:12:45.5E 31321269267102
\&   19:27:18.7N 027:59:22.1E 36195163180159
\&   15:12:40.2N 046:17:01.1E 40041499143669
\&   10:57:55.9N 064:12:52.8E 42906570007050
\&   06:43:07.3N 081:53:28.8E 44823504180200
\&   02:28:16.2N 099:24:54.5E 45813843358737
\&   01:46:36.0S 116:52:59.7E 45888525219677
.Ve
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fBGeoConvert\fR\|(1), \fBGeodSolve\fR\|(1).
.PP
An online version of this utility is available at
<https://geographiclib.sourceforge.io/cgi\-bin/RhumbSolve>.
.PP
An online version of this utility is available at
<https://geographiclib.sourceforge.io/cgi\-bin/RhumbSolve>.
.PP
This solution for rhumb line is described in C. F. F. Karney,
\&\fIThe area of rhumb polygons\fR,
Stud. Geophys. Geod. 68(3\-\-4), 99\-\-120 (2024);
DOI <https://doi.org/10.1007/s11200\-024\-0709\-z>.
.PP
The Wikipedia page, Rhumb line,
<https://en.wikipedia.org/wiki/Rhumb_line>.
.SH AUTHOR
.IX Header "AUTHOR"
\&\fBRhumbSolve\fR was written by Charles Karney.
.SH HISTORY
.IX Header "HISTORY"
\&\fBRhumbSolve\fR was added to GeographicLib,
<https://geographiclib.sourceforge.io>, in version 1.37 and
substantially rewritten in version 2.2.


NEWS
A reverse chronological list of changes to GeographicLib

For more information, see

    https://geographiclib.sourceforge.io/C++/doc/

The current version of the library is 2.7.

Changes between 2.7 (released 2025-11-06) and 2.6 versions:

   * Fix two bugs in classes for triaxial ellipsoids:
     + Triaxial::Geodesic3::Inverse sometimes threw an exception when
       both endpoints were on the major principal ellipse for very
       eccentric ellipsoids a > 2 c.
     + Triaxial::Cartesian3::carttocart2 sometimes returned the wrong
       result in the interior of the ellipsoid with Z = 0.

   * Cite the paper covering some of the triaxial classes Jacobi's
     solution for geodesics on a triaxial ellipsoid.

   * Minor documentation fixes.

Changes between 2.6 (released 2025-09-30) and 2.5.2 versions:

   * Support for triaxial ellipsoids. This is all placed in the
     GeographicLib::Triaxial namespace with the header files in the
     GeographicLib/Triaxial directory. This provides
     + Triaxial::Ellipsoid3 supports basic conversions of ellipsoidal
       coordinates.
     + Triaxial::GeodesicLine3 and Triaxial::Geodesic3 solve the direct
       and inverse geodesic problems on a triaxial ellipsoid. This
       implements Jacobi's solution for the direct problem. The inverse
       problem extends the method for solving the inverse problem on a
       biaxial ellipsoid.
     + Triaxial::Cartesian3 performs conversions between different
       systems of triaxial coordinates.
     + Triaxial::Conformal3 implements Jacobi's conformal projection for
       a triaxial ellipsoid. This allow a triaxial ellipsoid to be
       conformally mapped to a sphere or to any other ellipsoid.
     + Geod3Solve is a command-line utility offering access to
       Triaxial::Geodesic3.
     + Cart3Convert is a command-line utility offering access to
       Triaxial::Cartesian3.
     + Conformal3Proj is a command-line utility offering access to
       Triaxial::Conformal3.
     + Constants::Triaxial_Earth_a, Constants::Triaxial_Earth_b,
       Constants::Triaxial_Earth_c, and Constants::Triaxial_Earth_lon0
       specify a simple triaxial model for the Earth.
     + experimental::TriaxialGeodesicODE is provided to illustrate the
       solution of the direct geodesic problem by integrating the
       ordinary differential equations for the geodesic. Because this
       depends on Boost, this is not part of the GeographicLib library.
     + experimental/Geod3ODE.cpp a utility for doing direct geodesic
       calculations using experimental::TriaxialGeodesicODE.

   * To support the triaxial routines, the following classes have been
     added:
     + AngleT provides an accurate representation of angles of any
       magnitude. This might replace AuxAngle at some point.
     + Trigfun supports representing periodic functions as Fourier
       series with automatic selection of the number of terms to
       use. This offers a tiny subset of the functionality of the trig
       series in Chebfun. This might replace DST at some point.

   * C++17 is now required. Minimum version of Visual Studio supported
     is Visual Studio 15 2017.

   * Add Math::clamp.

   * Because experimental::TriaxialGeodesicODE depends on Boost, the
     CMake option USE_BOOST_FOR_EXAMPLES has been renamed
     USE_BOOST. This will cause find_package(Boost) to be called, but
     does not affect how the library or the tools are built.

Changes between 2.5.2 (released 2025-08-22) and 2.5.1 versions:

   * Fix potential buffer overflow in DMS.  Upgrading from earlier
     versions is strongly recommended.  If upgrading is not possible,
     avoid processing untrusted or malformed inputs and run programs
     linked to the library with restricted privileges.  Thanks to
     Rosario Matteo Grammatico for finding this vulnerability.

Changes between 2.5.1 (released 2025-08-20) and 2.5 versions:

   * Fix bug in geodesic inverse problem with very prolate
     ellipsoids. For f < -2, a meridional path was incorrectly
     returned in some cases for two points on the same meridian. This
     affected both the series and exact solutions. (Or course, the
     series solution is inaccurate for ellipsoids this eccentric.)

   * Minor tweak to Math::atan2d to improve accuracy.

   * Allow GEOGRAPHICLIB_PRECISION > 5 to allow use of mpreal with a
     fixed precision.

   * Document that cmake option -D BINDIR=OFF removes tools from the
     default all target for make.

   * WARNING: C++17 compliant compiler will be required for the next
     release. Support for builds with Visual Studio 14 (2015) will
     then stop.

Changes between 2.5 (released 2024-12-28) and 2.4 versions:

   * New magnetic models provided

     + World Magnetic Model 2025, wmm2025; this covers the period
       2025-2030.  This is now the model returned by
       MagneticModel::DefaultMagneticName and is default magnetic
       model for MagneticField (replacing wmm2020 which is only valid
       thru the end of 2024).
     + World Magnetic Model High Resolution 2025, wmmhr2025; this is a
       high-resolution version of wmm2025 including terms up to degree
       133 (as opposed to 12 for wmm2025).
     + International Geomagnetic Reference Field (14th generation),
       igrf14, which approximates the main magnetic field of the earth
       for the period 1900-2030.

   * Update CMake's cmake_minimum_required to 3.17.0 (to facilitate
     integration with Qt projects).

   * For GEOGRAPHICLIB_PRECISION > 2, readarray treats doubles as
     approximations to exact decimal constants.

Changes between 2.4 (released 2024-07-14) and 2.3 versions:

   * Math::sind, Math::sincosd, etc., now return accurate results for
     arguments = 30d, 45d, 60d, ...

   * Add Math::hypot3.

   * Add EllipticFunction::am.

   * Make constructor for AuxAngle explicit.

   * Include notes for plate carree areas.

   * C++14 compliant compiler required (this requirement tracks the

tools/RhumbSolve.cpp
/**
 * \file RhumbSolve.cpp
 * \brief Command line utility for rhumb line calculations
 *
 * Copyright (c) Charles Karney (2014-2023) <charles@karney.com> and licensed
 * under the MIT/X11 License.  For more information, see
 * https://geographiclib.sourceforge.io/
 *
 * See the <a href="RhumbSolve.1.html">man page</a> for usage information.
 **********************************************************************/

#include <iostream>
#include <string>
#include <sstream>
#include <fstream>
#include <cmath>
#include <limits>
#include <GeographicLib/Rhumb.hpp>
#include <GeographicLib/DMS.hpp>
#include <GeographicLib/Utility.hpp>

#if defined(_MSC_VER)
// Squelch warnings about constant conditional expressions and potentially
// uninitialized local variables
#  pragma warning (disable: 4127 4701)
#endif

#include "RhumbSolve.usage"
using namespace GeographicLib;
typedef Math::real real;

std::string LatLonString(real lat, real lon, int prec, bool dms, char dmssep,
                         bool longfirst) {
  using namespace GeographicLib;
  std::string
    latstr = dms ? DMS::Encode(lat, prec + 5, DMS::LATITUDE, dmssep) :
    DMS::Encode(lat, prec + 5, DMS::NUMBER),
    lonstr = dms ? DMS::Encode(lon, prec + 5, DMS::LONGITUDE, dmssep) :
    DMS::Encode(lon, prec + 5, DMS::NUMBER);
  return
    (longfirst ? lonstr : latstr) + " " + (longfirst ? latstr : lonstr);
}

std::string AzimuthString(real azi, int prec, bool dms, char dmssep) {
  return dms ? DMS::Encode(azi, prec + 5, DMS::AZIMUTH, dmssep) :
    DMS::Encode(azi, prec + 5, DMS::NUMBER);
}

int main(int argc, const char* const argv[]) {
  try {
    Utility::set_digits();
    bool linecalc = false, inverse = false, dms = false, exact = false,
      unroll = false, longfirst = false;
    real
      a = Constants::WGS84_a(),
      f = Constants::WGS84_f();
    real lat1, lon1, azi12 = Math::NaN(), lat2, lon2, s12, S12;
    int prec = 3;
    std::string istring, ifile, ofile, cdelim;
    char lsep = ';', dmssep = char(0);

    for (int m = 1; m < argc; ++m) {
      std::string arg(argv[m]);
      if (arg == "-i") {
        inverse = true;
        linecalc = false;
      } else if (arg == "-L") {
        inverse = false;
        linecalc = true;
        if (m + 3 >= argc) return usage(1, true);
        try {
          DMS::DecodeLatLon(std::string(argv[m + 1]), std::string(argv[m + 2]),
                            lat1, lon1, longfirst);
          azi12 = DMS::DecodeAzimuth(std::string(argv[m + 3]));
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding arguments of -L: " << e.what() << "\n";
          return 1;
        }
        m += 3;
      } else if (arg == "-e") {
        if (m + 2 >= argc) return usage(1, true);
        try {
          a = Utility::val<real>(std::string(argv[m + 1]));
          f = Utility::fract<real>(std::string(argv[m + 2]));
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding arguments of -e: " << e.what() << "\n";
          return 1;
        }
        m += 2;
      } else if (arg == "-u")
        unroll = true;
      else if (arg == "-d") {
        dms = true;
        dmssep = '\0';
      } else if (arg == "-:") {
        dms = true;
        dmssep = ':';
      } else if (arg == "-w")
        longfirst = !longfirst;
      else if (arg == "-p") {
        if (++m == argc) return usage(1, true);
        try {
          prec = Utility::val<int>(std::string(argv[m]));
        }
        catch (const std::exception&) {
          std::cerr << "Precision " << argv[m] << " is not a number\n";
          return 1;
        }
      } else if (arg == "-E")
        exact = true;
      else if (arg == "--input-string") {
        if (++m == argc) return usage(1, true);
        istring = argv[m];
      } else if (arg == "--input-file") {
        if (++m == argc) return usage(1, true);
        ifile = argv[m];
      } else if (arg == "--output-file") {
        if (++m == argc) return usage(1, true);
        ofile = argv[m];
      } else if (arg == "--line-separator") {
        if (++m == argc) return usage(1, true);
        if (std::string(argv[m]).size() != 1) {
          std::cerr << "Line separator must be a single character\n";
          return 1;
        }
        lsep = argv[m][0];
      } else if (arg == "--comment-delimiter") {
        if (++m == argc) return usage(1, true);
        cdelim = argv[m];
      } else if (arg == "--version") {
        std::cout << argv[0] << ": GeographicLib version "
                  << GEOGRAPHICLIB_VERSION_STRING << "\n";
        return 0;
      } else
        return usage(!(arg == "-h" || arg == "--help"), arg != "--help");
    }

    if (!ifile.empty() && !istring.empty()) {
      std::cerr << "Cannot specify --input-string and --input-file together\n";
      return 1;
    }
    if (ifile == "-") ifile.clear();
    std::ifstream infile;
    std::istringstream instring;
    if (!ifile.empty()) {
      infile.open(ifile.c_str());
      if (!infile.is_open()) {
        std::cerr << "Cannot open " << ifile << " for reading\n";
        return 1;
      }
    } else if (!istring.empty()) {
      std::string::size_type m = 0;
      while (true) {
        m = istring.find(lsep, m);
        if (m == std::string::npos)
          break;
        istring[m] = '\n';
      }
      instring.str(istring);
    }
    std::istream* input = !ifile.empty() ? &infile :
      (!istring.empty() ? &instring : &std::cin);

    std::ofstream outfile;
    if (ofile == "-") ofile.clear();
    if (!ofile.empty()) {
      outfile.open(ofile.c_str());
      if (!outfile.is_open()) {
        std::cerr << "Cannot open " << ofile << " for writing\n";
        return 1;
      }
    }
    std::ostream* output = !ofile.empty() ? &outfile : &std::cout;

    const Rhumb rh(a, f, exact);
    const RhumbLine rhl(linecalc ? rh.Line(lat1, lon1, azi12) :
                        rh.Line(0, 0, Math::qd));
    // Max precision = 10: 0.1 nm in distance, 10^-15 deg (= 0.11 nm),
    // 10^-11 sec (= 0.3 nm).
    prec = std::min(10 + Math::extra_digits(), std::max(0, prec));
    std::string s, eol, slat1, slon1, slat2, slon2, sazi, ss12, strc;
    std::istringstream str;
    int retval = 0;
    while (std::getline(*input, s)) {
      try {
        eol = "\n";
        if (!cdelim.empty()) {
          std::string::size_type m = s.find(cdelim);
          if (m != std::string::npos) {
            eol = " " + s.substr(m) + "\n";
            s = s.substr(0, m);
          }
        }
        str.clear(); str.str(s);
        if (linecalc) {
          if (!(str >> ss12))
            throw GeographicErr("Incomplete input: " + s);
          if (str >> strc)
            throw GeographicErr("Extraneous input: " + strc);
          s12 = Utility::val<real>(ss12);
          rhl.GenPosition(s12, Rhumb::ALL | (unroll ? Rhumb::LONG_UNROLL : 0),
                          lat2, lon2, S12);
          *output << LatLonString(lat2, lon2, prec, dms, dmssep, longfirst)
                  << " " << Utility::str(S12, std::max(prec-7, 0)) << eol;
        } else if (inverse) {
          if (!(str >> slat1 >> slon1 >> slat2 >> slon2))
            throw GeographicErr("Incomplete input: " + s);
          if (str >> strc)
            throw GeographicErr("Extraneous input: " + strc);
          DMS::DecodeLatLon(slat1, slon1, lat1, lon1, longfirst);
          DMS::DecodeLatLon(slat2, slon2, lat2, lon2, longfirst);
          rh.Inverse(lat1, lon1, lat2, lon2, s12, azi12, S12);
          *output << AzimuthString(azi12, prec, dms, dmssep) << " "
                  << Utility::str(s12, prec) << " "
                  << Utility::str(S12, std::max(prec-7, 0)) << eol;
        } else {                // direct
          if (!(str >> slat1 >> slon1 >> sazi >> ss12))
            throw GeographicErr("Incomplete input: " + s);
          if (str >> strc)
            throw GeographicErr("Extraneous input: " + strc);
          DMS::DecodeLatLon(slat1, slon1, lat1, lon1, longfirst);
          azi12 = DMS::DecodeAzimuth(sazi);
          s12 = Utility::val<real>(ss12);
          rh.GenDirect(lat1, lon1, azi12, s12,
                       Rhumb::ALL | (unroll ? Rhumb::LONG_UNROLL : 0),
                       lat2, lon2, S12);
          *output << LatLonString(lat2, lon2, prec, dms, dmssep, longfirst)
                  << " " << Utility::str(S12, std::max(prec-7, 0)) << eol;
        }
      }
      catch (const std::exception& e) {
        // Write error message cout so output lines match input lines
        *output << "ERROR: " << e.what() << "\n";
        retval = 1;
      }
    }
    return retval;
  }
  catch (const std::exception& e) {
    std::cerr << "Caught exception: " << e.what() << "\n";
    return 1;
  }
  catch (...) {
    std::cerr << "Caught unknown exception\n";
    return 1;
  }
}
